File: src/graphs/python/kruskal/src/graph.py
class Vertex:
    def __init__(self, value):
        self.value = value

    def __eq__(self, other):

        if isinstance(other, Vertex):
            return self.value == other.value

        return False

    def __lt__(self, other):

        if not isinstance(other, Vertex):
            raise NotImplementedError

        return self.value < other.value

    def __repr__(self):
        return str(self.value)

    def __hash__(self):
        return hash(self.value)


class Edge:
    def __init__(self, source, destination, distance):
        self.source = source
        self.destination = destination
        self.distance = distance

    def __eq__(self, other):

        if isinstance(other, Edge):
            return (
                self.source == other.source
                and self.destination == other.destination
                and self.distance == other.distance
            )

        return False

    def __lt__(self, other):

        if not isinstance(other, Edge):
            raise NotImplementedError

        return self.distance < other.distance

    def __repr__(self):
        return f"({self.source}, {self.destination}, {self.distance})"


class Graph:
    def __init__(self):
        self._adj_dict = dict()

    def add_edge(self, edge):

        if edge.source in self._adj_dict:
            self._adj_dict[edge.source].append(edge)

        else:
            self._adj_dict[edge.source] = [edge]

        if edge.destination not in self._adj_dict:
            self._adj_dict[edge.destination] = list()

    def add_vertex(self, vertex):
        if vertex not in self._adj_dict:
            self._adj_dict[vertex] = list()

    def vertices(self):
        return list(self._adj_dict.keys())

    def edges(self):
        return [elem for array in self._adj_dict.values() for elem in array if array]

    def edges_from_vertex(self, vertex):
        self._assert_graph_contains_vertex(vertex)
        return self._adj_dict[vertex]

    def contains(self, vertex):
        return vertex in self._adj_dict

    def connected(self, vertex_a, vertex_b):
        self._assert_graph_contains_vertex(vertex_a)
        self._assert_graph_contains_vertex(vertex_b)

        for edge in self._adj_dict[vertex_a]:
            if edge.destination == vertex_b:
                return True

        for edge in self._adj_dict[vertex_b]:
            if edge.destination == vertex_a:
                return True

        return False

    def empty(self):
        return len(self._adj_dict) == 0

    def size(self):
        return len(self._adj_dict)

    def __repr__(self):

        result = ""
        for vertex, edges in self._adj_dict.items():
            result += f"{vertex} —> "
            for edge in edges:
                result += edge
            result += "\n"

        return result

    def _assert_graph_contains_vertex(self, vertex):
        if vertex not in self._adj_dict:
            raise KeyError(f"Graph doesn't containt the {vertex} vertex!")

File: src/graphs/cpp/a_star/src/graph.cpp
#include "graph.h"
#include <algorithm>
#include <iterator>
#include <sstream>
#include <type_traits>

Point::Point(int _x, int _y) : x(_x), y(_y){};

bool Point::operator==(const Point &other) const {
  return x == other.x && y == other.y;
}

bool Point::operator!=(const Point &other) const {
  return x != other.x || y != other.y;
}

bool Point::operator<(const Point &other) const {
  return x < other.x && y < other.y;
}

template <class T> Vertex<T>::Vertex(T _value) : value(_value) {}

template <class T>
Vertex<T>::Vertex(const Vertex<T> &vertex) : value(vertex.value) {}

template <class T> bool Vertex<T>::operator==(const Vertex<T> &other) const {
  return value == other.value;
}

template <class T> bool Vertex<T>::operator!=(const Vertex<T> &other) const {
  return value != other.value;
}

template <class T> bool Vertex<T>::operator<(const Vertex<T> &other) const {
  return value < other.value;
}

template <class T>
size_t HashFunction<T>::operator()(const Vertex<T> &vertex) const {
  return std::hash<T>()(vertex.value);
}

template <>
size_t HashFunction<Point>::operator()(const Vertex<Point> &vertex) const {
  std::size_t h1 = std::hash<int>{}(vertex.value.x);
  std::size_t h2 = std::hash<int>{}(vertex.value.y);
  return h1 ^ (h2 << 1);
}

template <class T>
Edge<T>::Edge(Vertex<T> _source, Vertex<T> _destination, int _distance)
    : source(_source), destination(_destination), distance(_distance) {}

template <class T> bool Edge<T>::operator==(const Edge<T> &other) const {
  return source == other.source && destination == other.destination &&
         distance == other.distance;
}

template <class T> bool Edge<T>::operator!=(const Edge<T> &other) const {
  return source != other.source || destination != other.destination ||
         distance != other.distance;
}

template <class T> bool Edge<T>::operator<(const Edge<T> &other) const {
  return distance < other.distance;
}

template <class T> Graph<T>::Graph() {}

template <class T>
Graph<T>::Graph(
    std::unordered_map<Vertex<T>, std::vector<Edge<T>>, HashFunction<T>>
        adjcDict)
    : adjcDict(adjcDict) {}

template <class T> void Graph<T>::addEdge(Edge<T> edge) {
  auto source = edge.source;
  auto destination = edge.destination;

  if (contains(source))
    adjcDict[source].emplace_back(edge);

  else
    adjcDict[source] = std::vector<Edge<T>>{edge};

  if (!contains(destination))
    adjcDict[destination] = std::vector<Edge<T>>{};
}

template <class T>
void Graph<T>::addEdge(Vertex<T> source, Vertex<T> destination, int distance) {
  addEdge(Edge<T>(source, destination, distance));
}

template <class T> void Graph<T>::addVertex(Vertex<T> vertex) {
  if (!contains(vertex))
    adjcDict[vertex] = std::vector<Edge<T>>{};
}

template <class T> std::vector<Vertex<T>> Graph<T>::vertices() const {
  std::vector<Vertex<T>> result;

  for (const auto &[key, _] : adjcDict)
    result.emplace_back(key);

  return result;
}

template <class T>
std::vector<Edge<T>> Graph<T>::edgesFromVertex(const Vertex<T> vertex) const {
  if (!contains(vertex))
    throw std::invalid_argument("Graph<T> doesn't contain given vertex.");

  return adjcDict.at(vertex);
}

template <class T> std::vector<Edge<T>> Graph<T>::edges() const {
  std::vector<Edge<T>> result;

  for (auto [vertex, _edges] : adjcDict) {
    std::copy(_edges.begin(), _edges.end(), std::back_inserter(result));
  }

  return result;
}

template <class T> bool Graph<T>::contains(const Vertex<T> vertex) const {
  return adjcDict.find(vertex) != adjcDict.end();
}

template <class T>
bool Graph<T>::connected(const Vertex<T> source,
                         const Vertex<T> destination) const {
  if (!contains(source) || !contains(destination))
    return false;

  for (auto &edge : adjcDict.at(source)) {
    if (edge.destination == destination)
      return true;
  }

  return false;
}

template <class T> unsigned int Graph<T>::size() const {
  return adjcDict.size();
}

template <class T> bool Graph<T>::empty() const { return adjcDict.empty(); }

template class Vertex<int>;
template class Vertex<float>;
template class Vertex<double>;
template class Vertex<char>;
template class Vertex<Point>;

template class Edge<int>;
template class Edge<float>;
template class Edge<double>;
template class Edge<char>;
template class Edge<Point>;

template class Graph<int>;
template class Graph<float>;
template class Graph<double>;
template class Graph<char>;
template class Graph<Point>;
